File: src/AdventOfCode/Puzzles/Y2022/Day09.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2022;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2022/day/9</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2022, 09, "Rope Bridge", RequiresData = true)]
public sealed class Day09 : Puzzle
{
    /// <summary>
    /// Gets the number of positions that the tail of the rope visits at least once.
    /// </summary>
    public int PositionsVisited { get; private set; }

    /// <summary>
    /// Moves a rope using the specified moves and returns the
    /// number of positions that the tail of the rope visits at least once.
    /// </summary>
    /// <param name="moves">The moves to put the rope through.</param>
    /// <returns>
    /// The number of positions that the tail of the rope visits at least once.
    /// </returns>
    public static int Move(IList<string> moves)
    {
        var directions = Parse(moves);
        var rope = new Rope();

        var positions = new HashSet<Point>(moves.Count)
        {
            rope.Tail,
        };

        foreach (var direction in directions)
        {
            rope.Move(direction, (move) => positions.Add(move.Tail));
        }

        return positions.Count;

        static List<Size> Parse(IList<string> moves)
        {
            var result = new List<Size>(moves.Count);

            foreach (string move in moves)
            {
                char direction = move[0];
                int distance = Parse<int>(move[1..]);

                result.Add(direction switch
                {
                    'U' => new(0, distance),
                    'D' => new(0, -distance),
                    'L' => new(-distance, 0),
                    'R' => new(distance, 0),
                    _ => throw new PuzzleException($"Invalid direction '{direction}'."),
                });
            }

            return result;
        }
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(args);

        var moves = await ReadResourceAsLinesAsync();

        PositionsVisited = Move(moves);

        if (Verbose)
        {
            Logger.WriteLine("The tail of the rope visits {0} positions at least once", PositionsVisited);
        }

        return PuzzleResult.Create(PositionsVisited);
    }

    /// <summary>
    /// A class representing a rope with a length of 2 units.
    /// </summary>
    public sealed class Rope
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="Rope"/> class.
        /// </summary>
        /// <param name="head">The optional position of the head of the rope.</param>
        /// <param name="tail">The optional position of the tail of the rope.</param>
        public Rope(Point head = default, Point tail = default)
        {
            Head = head;
            Tail = tail;
        }

        /// <summary>
        /// Gets the position of the rope's head.
        /// </summary>
        public Point Head { get; private set; }

        /// <summary>
        /// Gets the position of the rope's tail.
        /// </summary>
        public Point Tail { get; private set; }

        /// <summary>
        /// Moves the rope in the specified direction.
        /// </summary>
        /// <param name="direction">The direction to move the rope in.</param>
        /// <param name="onMove">A delegate to invoke when the rope is moved.</param>
        public void Move(Size direction, Action<(Point Head, Point Tail)> onMove)
        {
            int magnitude = Math.Max(Math.Abs(direction.Width), Math.Abs(direction.Height));
            var unit = direction / magnitude;

            for (int i = 0; i < magnitude; i++)
            {
                var previous = Head;
                Head += unit;

                if (previous != Tail && Head != Tail)
                {
                    if (Math.Abs(Head.Y - Tail.Y) > 1 || Math.Abs(Head.X - Tail.X) > 1)
                    {
                        if (unit.Height == 0)
                        {
                            Tail += new Size(unit.Width, Head.Y - Tail.Y);
                        }
                        else
                        {
                            Tail += new Size(Head.X - Tail.X, unit.Height);
                        }

                        onMove((Head, Tail));
                    }
                }
            }
        }
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2022/Day09.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { repeat } from 'linq-to-typescript';
import { Point } from '../Point';
import { Puzzle } from '../Puzzle';
import { Size } from '../Size';
import { Puzzle2022 } from './Puzzle2022';

export class Day09 extends Puzzle2022 {
    positionsVisited2: number;
    positionsVisited10: number;

    override get name() {
        return 'Rope Bridge';
    }

    override get day() {
        return 9;
    }

    protected override get requiresData() {
        return true;
    }

    static move(moves: string[], knots: number): number {
        const parse = (moves: string[]) => {
            const result: Size[] = [];

            for (const move of moves) {
                const direction = move[0];
                const distance = Puzzle.parse(move.slice(1));

                let d: Size;

                switch (direction) {
                    case 'U':
                        d = new Size(0, distance);
                        break;
                    case 'D':
                        d = new Size(0, -distance);
                        break;
                    case 'L':
                        d = new Size(-distance, 0);
                        break;
                    case 'R':
                        d = new Size(distance, 0);
                        break;
                    default:
                        throw new Error(`Invalid direction '${direction}'.`);
                }

                result.push(d);
            }

            return result;
        };

        const directions = parse(moves);
        const rope = new Rope(repeat(Point.empty, knots).toArray());

        const positions = new Set<string>();
        positions.add(rope.tail.asString());

        for (const direction of directions) {
            rope.move(direction, (tail) => positions.add(tail.asString()));
        }

        return positions.size;
    }

    override solveCore(_: string[]) {
        const grid = this.readResourceAsLines();

        this.positionsVisited2 = Day09.move(grid, 2);
        this.positionsVisited10 = Day09.move(grid, 10);

        console.info(`The tail of the rope with two knots visits ${this.positionsVisited2} positions at least once.`);
        console.info(`The tail of the rope with ten knots visits ${this.positionsVisited10} positions at least once.`);

        return this.createResult([this.positionsVisited2, this.positionsVisited10]);
    }
}

class Rope {
    constructor(private readonly knots: Point[]) {}

    get count() {
        return this.knots.length;
    }

    get head() {
        return this.knots[0];
    }

    private set head(value: Point) {
        this.knots[0] = value;
    }

    get tail() {
        return this.knots[this.knots.length - 1];
    }

    move(direction: Size, onMoveTail: (tail: Point) => void) {
        const magnitude = Math.max(Math.abs(direction.width), Math.abs(direction.height));
        const unit = direction.divide(magnitude);

        for (let i = 0; i < magnitude; i++) {
            const previousTail = this.tail;
            this.head = Point.add(this.head, unit);

            for (let j = 1; j < this.count; j++) {
                const leader = this.knots[j - 1];
                let follower = this.knots[j];

                const deltaX = leader.x - follower.x;
                const deltaY = leader.y - follower.y;

                if (Math.abs(deltaX) > 1 || Math.abs(deltaY) > 1) {
                    follower = Point.add(follower, new Size(Math.sign(deltaX), Math.sign(deltaY)));
                }

                this.knots[j] = follower;
            }

            if (!previousTail.equals(this.tail)) {
                onMoveTail(this.tail);
            }
        }
    }
}
